drivers/counter/counter_nrfx_rtc.c

```diff
@@ -257,38 +324,124 @@ static int ppi_setup(struct device *dev)
-	task = nrfx_rtc_task_address_get(rtc, NRF_RTC_TASK_CLEAR);
+	evt_addr = nrf_rtc_event_address_get(rtc, evt);
+	task_addr = nrf_rtc_task_address_get(rtc, NRF_RTC_TASK_CLEAR);
 
 	result = nrfx_ppi_channel_alloc(&data->ppi_ch);
 	if (result != NRFX_SUCCESS) {
 		ERR("Failed to allocate PPI channel.");
 		return -ENODEV;
 	}
 
-	(void)nrfx_ppi_channel_assign(data->ppi_ch, evt, task);
+	(void)nrfx_ppi_channel_assign(data->ppi_ch, evt_addr, task_addr);
 	(void)nrfx_ppi_channel_enable(data->ppi_ch);
 #endif
 #endif /* CONFIG_COUNTER_RTC_WITH_PPI_WRAP */
 	return 0;
 }
 
-static int init_rtc(struct device *dev,
-		    const nrfx_rtc_config_t *config,
-		    nrfx_rtc_handler_t handler)
+static void ppi_free(struct device *dev)
+{
+#if CONFIG_COUNTER_RTC_WITH_PPI_WRAP
+	const struct counter_nrfx_config *nrfx_config = get_nrfx_config(dev);
+	u8_t ppi_ch = get_dev_data(dev)->ppi_ch;
+
+	if (!nrfx_config->use_ppi) {
+		return;
+	}
+#ifdef DPPI_PRESENT
+	NRF_RTC_Type *rtc = nrfx_config->rtc;
+	nrf_rtc_event_t evt = RTC_CHANNEL_EVENT_ADDR(chan);
+
+	(void)nrfx_dppi_channel_disable(ppi_ch);
+	nrf_rtc_subscribe_clear(rtc, NRF_RTC_TASK_CLEAR);
+	nrf_rtc_publish_clear(rtc, evt);
+	(void)nrfx_dppi_channel_free(ppi_ch);
+#else /* DPPI_PRESENT */
+	(void)nrfx_ppi_channel_free(ppi_ch);
+#endif
+#endif
+}
+
+/* Return true if counter must be cleared by the CPU. It is cleared
+ * automatically in case of max top value or PPI usage.
+ */
+static bool sw_wrap_required(struct device *dev)
+{
+	return (get_dev_data(dev)->top != COUNTER_MAX_TOP_VALUE)
+#if CONFIG_COUNTER_RTC_WITH_PPI_WRAP
+		    && !get_nrfx_config(dev)->use_ppi
+#endif
+		;
+}
+
+static int set_top_value(struct device *dev, const struct counter_top_cfg *cfg)
+{
+	const struct counter_nrfx_config *nrfx_config = get_nrfx_config(dev);
+	NRF_RTC_Type *rtc = nrfx_config->rtc;
+	struct counter_nrfx_data *dev_data = get_dev_data(dev);
+	bool custom_top = (dev_data->top != COUNTER_MAX_TOP_VALUE);
+	bool custom_new_top = (cfg->ticks != COUNTER_MAX_TOP_VALUE);
+	u32_t ch_len = counter_get_num_of_channels(dev) - (custom_top ? 1 : 0);
+	u8_t top_ch = counter_get_num_of_channels(dev) - 1; /* last channel */
+	int err = 0;
+
+	for (int i = 0; i < ch_len; i++) {
+		/* Overflow can be changed only when all alarms are
+		 * disables.
+		 */
+		if (nrfx_config->ch_data[i].callback) {
+			return -EBUSY;
+		}
+	}
+
+	if (custom_top && !custom_new_top) {
+		disable(dev, top_ch);
+		nrfx_config->ch_data[top_ch].callback = NULL;
+		ppi_free(dev);
+	}
+
+	nrf_rtc_task_trigger(rtc, NRF_RTC_TASK_CLEAR);
+
+	dev_data->top_cb = cfg->callback;
+	dev_data->top_user_data = cfg->user_data;
+	dev_data->top = cfg->ticks;
+
+	if (custom_new_top) {
+		int int_mask = RTC_CHANNEL_INT_MASK(top_ch);
+
+		err = ppi_setup(dev, top_ch);
+		if (err == 0) {
+			nrf_rtc_cc_set(rtc, top_ch, cfg->ticks);
+			/* mark channel as busy */
+			nrfx_config->ch_data[top_ch].callback = (void *)1;
```

Comment:
I'm afraid this is not a good idea. If one channel is needed for handling the top value, it should not be made available for API users. With the above implementation, the driver reports via `counter_get_num_of_channels` that certain number of channels is available for use, but after the TOP value is set to something different than what is reported by `counter_get_max_top_value`, apparently one of these channels becomes unusable (permanently busy).

Reply:
on the other hand, it is very likely that user will use default top value and in that case it's a waste. Alternatively, i can add compile time option per instance to enable custom top value or not. what do you think?

Reply:
I agree that it may be confusing but as i mentioned we might need this one more channel badly when counter will be used as system timer and we would like to share it with bluetooth to free up one RTC. I think that best would be to add `fixed_top` flag to instance configuration and do it in similar way as `ppi_wrap` so that if all instances are using fixed top value then support for custom top is removed. That can also simplify functions for ticks operations (add, sub). However, i started this and it will be more changes so would propose to do it as another PR and get that one in. Do you agree?

Reply:
@anangl I've created issue for that #19569.

Reply:
The idea of `fixed_top` flag seems to be reasonable but I'm not convinced that this is the right way to proceed: to introduce this confusing allocation of one channel in this PR and later, potentially, remove it with another one, calling it an enhancement.

Reply:
ok, i've done following: added fixed_top flag per instance. Also added a kconfig flag which is set if all instances have fixed_top set. If yes much of the counter code is removed (significant driver code size reduction 1,5k -> 1k). Added dedicated test suite to validate correct behavior when fixed_top is set for all instances.

Additionally, enabled ppi-wrap on one instance in the test (it revealed that ppi-wrap was broken) added ppi-wrap fix.